### PR TITLE
Native inbox Course picker

### PR DESCRIPTION
**Pull request by domonkosadam**
refs: MBL-14854
affects: Teacher, Student
release note: none
test plan:
- enable student inbox experimental feature
- send a new message
- open course picker

**Changes**:

- Native Course picker dialog
- Course colours added to list
- Groups added to list

**What's not included**:

- Address book currently not works with groups (Will be fixed in ticket MBL-14855)

## Screenshots

<table>
<tr><th>Before</th><th>After</th></tr>
<tr>
<td></td>
<td></td>
</tr>
</table>

Native inbox Course picker

refs: MBL-14854
affects: Student, Teacher
release note: none

## Checklist

- [x] A11y checked
- [x] Tested on phone
- [x] Tested on tablet
- [x] Tested in dark mode
- [x] Tested in light mode

### CODE REVIEW

**Comment by inst-danger** [comment]
<details><summary>Parent Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Teacher Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<details><summary>Student Build QR Code:</summary>

</details>

**Comment by inst-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add a release note. If no release note is wanted, use `none`. Example: `release note: Fixed a bug that prevented users from enjoying the app.`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add which apps this change affects. Example: `affects: Teacher, Student` or `affects: none`
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  Please add a reference to a jira ticket. For example: `refs: MBL-10023`
  </td>
    </tr>
  </tbody>
</table>


<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Warnings</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:warning:</td>
      <td>One or more files are below the minimum test coverage 50%</td>
    </tr>
  </tbody>
</table>


[MBL-14854](https://instructure.atlassian.net/browse/MBL-14854)

Coverage | New % | Master % | Delta
-------- | ----- | -------- | -----
**Canvas iOS** | 91.58% | 91.82% | -0.24%
Core/Core/Inbox/ComposeMessage/Model/RecipientContext.swift | 0% | -- | --
Core/Core/Inbox/InboxCoursePicker/ViewModel/InboxCoursePickerViewModel.swift | 0% | -- | --
Core/Core/Inbox/ComposeMessage/Model/ComposeMessageInteractorLive.swift | 0% | 0% | 0%
Core/Core/Inbox/ComposeMessage/ViewModel/ComposeMessageViewModel.swift | 38.17% | 33.97% | 4.2%
Core/Core/Inbox/InboxCoursePicker/Model/InboxCoursePickerInteractorLive.swift | 0% | -- | --
Core/Core/Inbox/Addressbook/ViewModel/AddressbookViewModel.swift | 0% | 0% | 0%
Core/Core/Inbox/Addressbook/Model/AddressbookInteractorLive.swift | 0% | 0% | 0%
Core/Core/Inbox/ComposeMessage/Model/MessageParameters.swift | 0% | 0% | 0%
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 25625e4e7ecc03e8d9820bd3809df9ebdb217115
</p>

File: Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    private func headerView(_ header: String) -> some View {
+        VStack(spacing: 0) {
+            separator
+            Text(header)
+                .font(.regular14)
+                .frame(maxWidth: .infinity, alignment: .leading)
+                .foregroundStyle(Color.textDark)
+                .padding(.horizontal, 12)
+                .padding(.vertical, 16)
+                .background(Color.backgroundLight)
```
Comment (vargaat) [inline Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift:92]: Marking this text as header for voiceover would make the purpose of this text more obvious.

File: Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            }
+            .padding(.vertical, 16)
+            separator
+        }
+        .contentShape(Rectangle())
+        .onTapGesture {
+            let recipientContext = RecipientContext(group)
+            viewModel.selectedRecipientContext = recipientContext
+            viewModel.didSelect?(recipientContext)
+        }
```
Comment (vargaat) [inline Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift:160]: My comments on the course row apply here as well.

File: Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+    private func groups(groups: [Group]) -> some View {
+        VStack(spacing: 0) {
+            headerView(NSLocalizedString("Groups", bundle: .core, comment: ""))
+            ForEach(groups, id: \.id) { group in
+                groupRow(group)
+            }
+        }
+    }
+
+    private func headerView(_ header: String) -> some View {
```
Comment (vargaat) [inline Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift:83]: Check if SwittUI-s built-in Section works with scrollviews or not. That automatically makes the header view passed to it sticky.

**Review by vargaat — CHANGES_REQUESTED** [review]
Good job in overall!

- I noticed that when selecting a long course name it looks weird that it's center aligned on the compose screen, we could change that to a leading alignment. 
- A minor UI glitch that when selecting a row the height of the row increases to make space for the checkmark. We could either reduce the checkmark's size or add a minimum height to the row to reserve the space for the checkmark.
- I would change the background color of the view to the one we have for the section headers to avoid the last group cell blending into the background without visual separation.

File: Core/Core/Inbox/InboxCoursePicker/ViewModel/InboxCoursePickerViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    // MARK: - Input / Output
+    @Published public var selectedRecipientContext: RecipientContext?
+
+    // MARK: - Inputs
+    public private(set) var dismissViewDidTrigger = PassthroughSubject<Void, Never>()
+    var didSelect: ((RecipientContext) -> Void)?
+
+    // MARK: - Private
+    private var subscriptions = Set<AnyCancellable>()
```
Comment (vargaat) [inline Core/Core/Inbox/InboxCoursePicker/ViewModel/InboxCoursePickerViewModel.swift:36]: Unused.

File: Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+                    separator
+                            Text("Courses")
+                                .font(.regular14)
+                                .frame(maxWidth: .infinity, alignment: .leading)
+                                .foregroundStyle(Color.textDark)
+                                .padding(.horizontal, 12)
+                                .padding(.vertical, 16)
+                                .background(Color.backgroundLight)
+                                .accessibilityHeading(.h1)
+                    separator
```
Comment (vargaat) [inline Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift:75]: Indentation seems off. We could extract maybe the header builder to make this part smaller just like you did with the courseRow method.

File: Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+                        VStack(spacing: 0) {
+                            Text("Groups")
+                                .font(.regular14)
+                                .frame(maxWidth: .infinity, alignment: .leading)
+                                .foregroundStyle(Color.textDark)
+                                .padding(.horizontal, 12)
+                                .padding(.vertical, 16)
+                                .background(Color.backgroundLight)
+                                .accessibilityHeading(.h1)
+                    separator
```
Comment (vargaat) [inline Core/Core/Inbox/InboxCoursePicker/View/InboxCoursePickerView.swift:99]: Indentation is off.

**Review by vargaat — CHANGES_REQUESTED** [review]
Thank you for the fixes. There are a few items left from my previous review and I noticed these during this review cycle.
- Seems like we lost the left alignment in the context selector screen.
- Groups don't load recipients and the error page seems a little strange in dark mode.
- A random improvement idea to turn on auto capitalization for the subject and the message fields for sentences.

**Review by szabinst — CHANGES_REQUESTED** [review]
I think we could improve the following three things: 
- When navigating back from the Course picker to the New Message screen, the "New Message" title is misplaced and there's an animation where it moves back to its original place. 
- I think we should include a loading indicator to the Course Picker screen, so we don't have a blank screen while the courses are loading. 

https://github.com/instructure/canvas-ios/assets/110813321/4873e273-9d9c-41bf-825b-0f3f3b0e78f0)

The error screen for no recipients use case, looks a bit unpolished. 

IMG_0130

**Comment by domonkosadam** [comment]
The address book will be fixed in the following ticket.

**Review by szabinst — APPROVED** [review]
QA + 1

**Review by vargaat — APPROVED** [review]

**Pull request merged**